modules/apigee_edge_apidocs/tests/src/Functional/ApiDocsAccessTest.php

```diff
@@ -0,0 +1,418 @@
+<?php
+
+/**
+ * Copyright 2018 Google Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU General Public License version 2 as published by the
+ * Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+ * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public
+ * License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; if not, write to the Free Software Foundation, Inc., 51
+ * Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+
+namespace Drupal\Tests\apigee_edge_apidocs\Functional;
+
+use Drupal\apigee_edge_apidocs\Entity\ApiDoc;
+use Drupal\Tests\BrowserTestBase;
+
+/**
+ * Tests the ApiDoc term access permissions.
+ *
+ * @group apigee_edge
+ * @group apigee_edge_apidocs
+ */
+class ApiDocsAccessTest extends BrowserTestBase {
```

Comment:
@peterserfozo
* Please describe why fewer test methods within a test class are better in Drupal 8 than more. 
* Please demonstrate to Chris how entity access can be tested in a better way in Drupal 8.

Reply:
In our case, it's better to use fewer test methods in a test class because Drupal is reinstalled before every test method execution in functional tests. Of course, creating well-separated test cases is a PHPUnit best practice but in Drupal 8 we should create fewer because of the heavy system setup. And if there are time-consuming operations (for example API calls) in `setUp()` then it's better to call it once because `setUp()` is called before executing test methods. On the other hand, our module is really complex, we use different static and persistent cache layers, persistent data, etc. It means if we create separated test cases then we may hide bugs because of the purged DB. Anyway, writing well-structured code is important, you can create a similar solution like this one: https://github.com/apigee/apigee-edge-drupal/blob/8.x-1.x/tests/src/Functional/QueryTest.php#L126

The best approach to test permissions and access system is using a permission matrix: a two-dimensional array is enough to define which routes are accessible to the different permissions/roles then the test loops through this array, sets the roles of the user and validates whether the user should have access to a route or not in every step. To test entity access you can use the entity's route provider for collecting the routes defined by the entity. A good example: https://github.com/apigee/apigee-edge-drupal/blob/237ff6fbd5c83dd3b6efffafff82ff20f7dd26da/modules/apigee_edge_teams/tests/src/Functional/AccessTest.php